src/dataset.js

```diff
@@ -205,6 +208,28 @@ function Dataset(bigQuery, id) {
 
   this.bigQuery = bigQuery;
 
+  /**
+   * @name Dataset#metadata
+   * @type {object}
+   */
+  this.metadata = options || {};
+
+  /*!
```

Comment:
Now that we're not allowing lazy-setting of `location`, what are the scenarios that this lazy-getting would be used?

Reply:
I found this to be conveinent when building requests so I didn't have to do something like this all over the place.

```js
var qs = {location: this.metadata && this.metadata.location};

// or for job requests

if (this.metadata && this.metadata.jobReference) {
  qs = {location: this.metadata.jobReference.location};
}
```

Because the location would be in different spots depending on the object, normalizing the location seems like a pretty convenient thing to do for the user on the off chance they wanted to reference it.

Reply:
I'm cool with the aliasing, but if we're always getting the value up front from the user, couldn't we just do:

```js
this.location = options.location;
```

The comment talks about how we cache it if the location comes back in an apiResponse, but I think we're not relying on that anymore.

Reply:
Ah, I understand. I think there's still a need for this to allow the user to do something like so

```js
const dataset = bigquery.dataset('name');

await dataset.get();
const [job] = await dataset.createQueryJob({...});
```

Reply:
Per the other conversation about hierarchy, I don't think we should support this behavior, since we demonstrate in the examples and documentation that `location` should be set on the instance.

The user would have to accidentally fall into the pattern in your example above, unless there's a utility in fetching the location of a dataset from just a name.

But that has me wondering-- wouldn't `dataset.get()` fail if `name` is in a location other than the upstream API's default location? (sub-question-- can the same dataset name be used in two locations?)

Reply:
You're right, it would fail. Using whatever location came back in the `apiResponse` was in the pre-reqs, so I would prefer that @tswast or @jjolma make this decision.

Reply:
I went back to read the doc:

> - Extract the job’s location from any responses containing a JobReference.
> - Include a job’s location for requests operating on a fetched job. For example, if the client does a jobs.list and then does a jobs.cancel on one of the jobs, the cancel should include the job’s location.

For the methods that return a `JobReference`, we don't need to extract the location, because the user provided the location up-front. The only case I can think of (correct me if I'm wrong), where we don't know the location is `getJobs()`. Thus, I believe we can accomplish both requirements by changing `getJobs()`:

```js
var jobs = (resp.jobs || []).map(function(jobObject) {
  var job = that.job(jobObject.jobReference.jobId, {
    location: jobObject.jobReference.location,
  });
  job.metadata = jobObject;
  return job;
});
```

Reply:
I suppose that's fair, we'd probably want to do the same thing for `getDatasets` as well. However I think I'm struggling to understand why this is superior to lazy getting. I could be wrong, but this feels like bike shedding.

Reply:
Since we can avoid it, the lazy-getting creates a confusing point in the code, that doesn't have a purpose. Leaving leaks of hidden behavior and support for un-documented use cases makes the code hard to maintain. I just figure we should make the change as deliberate and simple as possible, so as to make it easier to understand/refactor some day.